torch/distributed/launch.py

```diff
@@ -242,6 +258,15 @@ def main():
             # create the relative directory
             os.mkdir(os.path.join(os.getcwd(), args.logdir))
 
+    if args.dist_url:
+        url_obj = urlparse(args.dist_url)
+        if url_obj.scheme == "file":
+            if args.nnodes > 1:
+                raise ValueError("FileStore only supports one node")
+            else:
+                check_filename(urllib.request.url2pathname(url_obj.path))
+                current_env["INIT_METHOD"] = args.dist_url
```

Comment:
Maybe I'm missing something here, but where exactly is the the "INIT_METHOD" env used?

Reply:
@pritamdamania87 , thank your reply. The way of  "INIT_METHOD" used is same as MASTER_ADDR and MASTER_PORT. The usage is in the example https://github.com/pytorch/examples/pull/845/files

Reply:
@mszhanyi I see, although in https://github.com/pytorch/examples/pull/845/files it seems like the env variable is only being used by an example script. What I meant is that I don't see any core components of DDP/c10d using this environment variable.

Reply:
@pritamdamania87 . Thank your reminder.  Searched the repo, I think the reasons are
1. For DDP, the most popular way is using MASTER_ADDR and MASTER_PORT. At first, I wanted to use dist_url/init_method to cover master_addr/master_port. But I'm afraid this requirement doesn't worth such a big change. Master_ADDR/MASTER_PORT should be the default option for users, so I don't set the default value for the init_method.
2. My purpose is that we should not encourage the user define the filepath of FileStore in the train.py and one option in launch.py should be provided to the FileStore users
3. In DDP/c10d, we use [TEMPFILE](https://github.com/pytorch/pytorch/blob/22a34bcf4e5eaa348f0117c414c3dd760ec64b13/c10/util/tempfile.h) to test FileStore. So, they don't need the environment variable.
Please corrects me if I'm wrong. Is there any better solution?

Reply:
My concern here is that the launcher script can be used with any arbitrary script and not just with the script in https://github.com/pytorch/examples/pull/845/files. As a result, if a user specifies `--dist_url` with their custom script, this would have no effect at all and it would be an inconsistency.

If the overall goal here is to check whether the path for FileStore is valid, we should probably check it in the constructor of FileStore itself?

Reply:
The FileStore ctor is called in each training process so it couldn't solve the long time hang issue if we add check in the ctor FileStore.

I think if users set the init_method in the API of init_process_group , it's their responsibility to make sure the url is right.
For Maste_addr/Master_port in launcher.py, they could be invalid as well if users set the init_method as tcp://xxxx in the API of init_process_group. 

If users follow our instruction/example to start distributed training, it'd better to raise the error quickly if they input some incorrect value.

Reply:
> The FileStore ctor is called in each training process so it couldn't solve the long time hang issue if we add check in the ctor FileStore.

Hey @mszhanyi, could you please elaborate on this concern? 


https://github.com/pytorch/pytorch/blob/e3542d2c12d8aaaccf8a53873e480c20dc6b7338/torch/lib/c10d/FileStore.cpp#L267-L278

IIUC, @pritamdamania was referring to adding a check to the above function, and error out if the file does not exist. Since the launch script will pass the same path to all processes, I would assume all processes should crash if the path is invalid, so there will be no hang? Or are you concerned about the case that the file exist on some node but not all, so that some process would crash but others would hang?

Reply:
@mrshenli, thank your feedback. 
The key point is we have to try creating the file to check if the filename is correct.
To avoid race among processes to create the file if the file doesn't exist #13750, The timeout is added.
https://github.com/pytorch/pytorch/blob/8177f63c91f7c6343859d12c6818e129acaa974d/torch/lib/c10d/FileStore.cpp#L147-L158
So if the check is added in FileStore ctor, the hang will still exist or there's the regression of #13750.
Please correct me if I'm wrong.

Reply:
Maybe the right thing to do here would be to replace the `while(true)` loop in FileStore with a loop that only retries creating the file 3-5 times. This way we wouldn't wait for the entire timeout duration if there is bad file path and at the same time it would address the race issue mentioned in  #13750